cpp/src/arrow/compute/expression.cc

```diff
@@ -1242,33 +1242,103 @@ struct Inequality {
                             /*insert_implicit_casts=*/false, &exec_context);
   }
 
+  /// Simplify an is_in predicate against this inequality as a guarantee.
+  Result<Expression> SimplifyIsIn(Expression expr) {
+    const auto& guarantee = *this;
+    auto call = expr.call();
+    auto options = checked_pointer_cast<SetLookupOptions>(call->options);
+
+    auto value_set = options->value_set.make_array();
+    if (!value_set) return expr;
+    if (value_set->length() == 0) return literal(false);
+
+    if (!options->sorted_and_deduped) return expr;
+
+    // For now, only simplify when the guarantee is non-nullable.
```

Comment:
🤔 why is the design here?

Reply:
Suppose the value set is sorted such that nulls are placed at the end. Then simplifying on a `less` or `less_equal` predicate wouldn't work as slicing the front half of the value set array would drop the null values.

To support this, I think we'd need the `is_in` predicate itself to "know" whether it contains a null separately from storing nulls in the value set, perhaps by keeping track using an additional flag in `SetLookupOptions`.

Reply:
```suggestion
    // Null-matching behavior is complex and reduces the chances of
    // reduction of `is_in` calls to a single literal for every possible
    // input, so we abort the simplification if nulls are possible in the
    // input.
```